File: src/main/scala/com/kotori316/fluidtank/network/FluidCacheMessage.java
package com.kotori316.fluidtank.network;

import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;

import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Registry;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.resources.ResourceKey;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.level.Level;

import com.kotori316.fluidtank.FluidTank;
import com.kotori316.fluidtank.fluids.FluidAmount;
import com.kotori316.fluidtank.tiles.CATTile;

/**
 * To Client Only.
 */
public class FluidCacheMessage implements IMessage<FluidCacheMessage> {
    public static final ResourceLocation NAME = new ResourceLocation(FluidTank.modID, "fluid_cache_message");
    private final ResourceKey<Level> dimensionId;
    private final BlockPos pos;
    private final List<FluidAmount> amounts;

    public FluidCacheMessage(CATTile tile) {
        pos = tile.getBlockPos();
        dimensionId = Optional.ofNullable(tile.getLevel()).map(Level::dimension).orElse(Level.OVERWORLD);
        amounts = tile.fluidAmountList();
    }

    public FluidCacheMessage(FriendlyByteBuf buffer) {
        pos = buffer.readBlockPos();
        dimensionId = ResourceKey.create(Registry.DIMENSION_REGISTRY, buffer.readResourceLocation());
        int size = buffer.readInt();
        amounts = Stream.generate(buffer::readNbt).limit(size)
            .map(FluidAmount::fromNBT)
            .toList();
    }

    @Override
    public void writeToBuffer(FriendlyByteBuf buffer) {
        buffer.writeBlockPos(pos).writeResourceLocation(dimensionId.location());
        buffer.writeInt(amounts.size());
        amounts.forEach(a -> buffer.writeNbt(a.write(new CompoundTag())));
    }

    @Override
    public ResourceLocation getIdentifier() {
        return NAME;
    }

    @Environment(EnvType.CLIENT)
    static class HandlerHolder {
        static final ClientPlayNetworking.PlayChannelHandler HANDLER = (client, handler, buf, responseSender) -> {
            var message = new FluidCacheMessage(buf);
            var world = client.level;
            if (world != null && world.dimension().equals(message.dimensionId)) {
                client.execute(() -> {
                    if (world.getBlockEntity(message.pos) instanceof CATTile tile) {
                        tile.fluidCache = message.amounts;
                    }
                });
            }
        };
    }
}

File: src/main/scala/com/kotori316/fluidtank/network/FluidCacheMessage.scala
package com.kotori316.fluidtank.network

import java.util.function.Supplier

import com.kotori316.fluidtank.FluidTank
import com.kotori316.fluidtank.fluids.FluidAmount
import com.kotori316.fluidtank.tiles.CATTile
import net.minecraft.core.{BlockPos, Registry}
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.resources.ResourceKey
import net.minecraft.world.level.Level
import net.minecraftforge.network.NetworkEvent

/**
 * To client only.
 */
case class FluidCacheMessage(dimensionId: ResourceKey[Level],
                             pos: BlockPos,
                             amounts: Seq[FluidAmount]) {
  def this(tile: CATTile) = {
    this(
      Option(tile.getLevel).map(_.dimension()).getOrElse(Level.OVERWORLD),
      tile.getBlockPos,
      tile.fluidAmountList
    )
  }

  def write(buffer: FriendlyByteBuf): Unit = {
    buffer.writeBlockPos(pos).writeResourceLocation(dimensionId.location)
    buffer.writeInt(amounts.size)
    amounts.foreach(a => buffer.writeNbt(a.write(new CompoundTag)))
  }

  def onReceive(ctx: Supplier[NetworkEvent.Context]): Unit = {
    FluidTank.proxy.getLevel(ctx.get).filter(_.dimension.equals(dimensionId))
      .map(_.getBlockEntity(pos))
      .collect { case c: CATTile => c }
      .foreach(tile => ctx.get.enqueueWork(() => tile.fluidCache = this.amounts))
    ctx.get.setPacketHandled(true)
  }
}

object FluidCacheMessage {
  def apply(buffer: FriendlyByteBuf): FluidCacheMessage = {
    val pos = buffer.readBlockPos
    val dimensionId = ResourceKey.create(Registry.DIMENSION_REGISTRY, buffer.readResourceLocation)
    val size = buffer.readInt
    val amounts = Seq.fill(size)(buffer.readNbt()).map(FluidAmount.fromNBT)
    new FluidCacheMessage(dimensionId, pos, amounts)
  }
}
